app/services/activitypub/fetch_remote_status_service.rb

```diff
@@ -80,4 +80,20 @@ def expected_activity_type?
   def expected_object_type?
     equals_or_includes_any?(@json['type'], ActivityPub::Activity::Create::SUPPORTED_TYPES + ActivityPub::Activity::Create::CONVERTED_TYPES)
   end
+
+  def fetch_status(uri, id_is_known, on_behalf_of = nil)
+    begin
+      fetch_resource(uri, id_is_known, on_behalf_of, raise_on_error: true)
+    rescue Mastodon::UnexpectedResponseError => e
+      return unless e.response.code == 404
+
+      # If this is a 404 from a status from an account that has no local followers, delete it
+      existing_status = Status.find_by(uri: uri)
+      if !existing_status.nil? && existing_status.unsubscribed?
```

Comment:
The `unsubscribed?` method may not be the correct condition here. We'd want to exclude private posts from this instead.

Reply:
ah yes, missed that. will add that & tests for not removing private posts that should correctly return an 404.

just checking - are you saying we should add an additional condition to exclude followers-only & mentioned-only statuses? or use that instead of `unsubscribed`? 

the `unsubscribed` check is to protect against a server being offline momentarily causing us to remove the post erroneously - if we were "subscribed" we would have received the deletion activity. the "orphaned"/"unsubscribed" case where we hold on to posts that have been deleted on remote server is the main reason this was requested, so if we want to take it out here, we may want to remove it entirely, since it is only used here and in `FetchAllRepliesService` which gathers unsubscribed statuses for this purpose.

Reply:
correctly failing tests for not removing nondistributable posts (oops, looks like test runs are cancelled if another push happens too quickly, these can be re-run to confirm if needed): https://github.com/mastodon/mastodon/pull/32615/commits/e5aada815ea42e1261f8af2ee8d961f978861ce5

fixed behavior, no longer removing nondistributable posts after a 404: https://github.com/mastodon/mastodon/pull/32615/commits/a050e0f97963790ef8e76726b06e71536bb91dda